File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/options/OptionsModule.java
package com.greentree.engine.moon.base.options;

import com.greentree.engine.moon.assets.key.ResourceAssetKey;
import com.greentree.engine.moon.base.AssetManagerProperty;
import com.greentree.engine.moon.base.property.modules.CreateProperty;
import com.greentree.engine.moon.base.property.modules.ReadProperty;
import com.greentree.engine.moon.modules.LaunchModule;
import com.greentree.engine.moon.modules.property.EngineProperties;

import java.util.Properties;

import static com.greentree.commons.reflection.info.TypeInfoBuilder.getTypeInfo;

public class OptionsModule implements LaunchModule {

    @ReadProperty(AssetManagerProperty.class)
    @CreateProperty(OptionsProperty.class)
    @Override
    public void launch(EngineProperties properties) {
        var manager = properties.get(AssetManagerProperty.class).manager;
        var ps = manager.load(getTypeInfo(Properties.class), new ResourceAssetKey("options.properties")).getValue();
        properties.add(new OptionsProperty(new PropertiesOptionsProvider(ps)));
    }

    private record PropertiesOptionsProvider(Properties properties) implements OptionsProvider {

        @Override
        public String get(String name) {
            return properties.getProperty(name);
        }

    }

}

File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/options/OptionsModule.kt
package com.greentree.engine.moon.base.options

import com.greentree.engine.moon.assets.loader.load
import com.greentree.engine.moon.base.AssetManagerProperty
import com.greentree.engine.moon.base.property.modules.CreateProperty
import com.greentree.engine.moon.base.property.modules.ReadProperty
import com.greentree.engine.moon.modules.LaunchModule
import com.greentree.engine.moon.modules.property.EngineProperties
import java.util.*

class OptionsModule : LaunchModule {

	@ReadProperty(AssetManagerProperty::class)
	@CreateProperty(OptionsProperty::class)
	override fun launch(properties: EngineProperties) {
		val manager = properties.get(AssetManagerProperty::class.java).manager
		val ps = manager.load<Properties>("options.properties")
		properties.add(OptionsProperty(PropertiesOptionsProvider(ps)))
	}

	private data class PropertiesOptionsProvider(val properties: Properties) : OptionsProvider {

		override fun get(name: String): String {
			return properties.getProperty(name)
		}
	}
}
